### PR TITLE
Enhancement: Small Improvement for Interactively Overriding Local Testing Harness Env Vars 

### PR DESCRIPTION
**Pull request by tzaffi**
Should I keep these changes around and merge `harness_overwrite.py`? This tinkering work grew out of an investigation of the downstream effects of this go-algorand PR: https://github.com/algorand/go-algorand/pull/4951

### Arguments For and Against Merging the PR
- [ ] DO NOT MERGE
  - [ ] This violates the goal of keeping `up.sh` as similar as possible across all 4 SDK's
  - [ ] Even if we decide that this _should_ exist in all 4 SDK's, it's written in Python, and we can't impose installing Python across the other SDK's
  - [ ] It's not useful enough to justify the future maintenance costs
- [ ] MERGE
  - [ ] It's okay to have small "violations" of the uniformity goal, if there is a clear upside
  - [ ] Anything that eases the investigation of downstream implications for changes in `go-algorand` -_even a little_- is worthwhile
  - [ ] Python is so ubiquitous for shell scripting; we could impose requiring it in all the SDK's - imagine how much easier life would be if we could dispense with shell scripting in favor of python scripting?

## Q: What does `harness_overwrite.py` do?
It makes it slightly easier _locally_ to mod environments that the test harness needs.  In particular

1. set `INTERACTIVE_TESTING_ENVIRONMENT=1` in `.test-env`
2. run `make harness` and follow the prompts.

### here are what the prompts look like

```sh
Which of the following env vars do you want to modify?

VERBOSE_HARNESS,TYPE,ALGOD_CHANNEL,ALGOD_URL,ALGOD_BRANCH,ALGOD_SHA,NETWORK_TEMPLATE,NETWORK_NUM_ROUNDS,INDEXER_URL,NODE_ARCHIVAL,INDEXER_BRANCH,INDEXER_SHA,SANDBOX_URL,SANDBOX_BRANCH,LOCAL_SANDBOX_DIR,ALGOD_CONTAINER,KMD_PORT,ALGOD_PORT,INDEXER_CONTAINER,INDEXER_PORT,POSTGRES_CONTAINER,POSTGRES_PORT

    (skip for ALL, or provide comma separated)
    
    A typical choice is:
TYPE,ALGOD_URL,ALGOD_BRANCH

CHOICES:
```
Then I entered the "typical choice" and proceeded with the next prompts:
```
...

CHOICES:
TYPE,ALGOD_URL,ALGOD_BRANCH
__________________________________________________
Please provide env variable overrides (skip to keep defaults)
TYPE (default `"channel"`):source
ALGOD_URL (default `"https://github.com/algorand/go-algorand"`):https://github.com/tzaffi/go-algorand
ALGOD_BRANCH (default `"master"`):improve-some-goal-errors
__________________________________________________
OVERWRITING
new_env[TYPE]=source (PREVIOUS: env[TYPE]="channel")
new_env[ALGOD_URL]=https://github.com/tzaffi/go-algorand (PREVIOUS: env[ALGOD_URL]="https://github.com/algorand/go-algorand")
new_env[ALGOD_BRANCH]=improve-some-goal-errors (PREVIOUS: env[ALGOD_BRANCH]="master")
__________________________________________________
...
```

## Q: What are the implications for `harness_overwrite.py` in C.I.?

As long as one remembers not to merge in `INTERACTIVE_TESTING_ENVIRONMENT=1` in `.test-env`,
C.I. will continue running as before. In order to run the same override environment in C.I., you
need to take the following steps:

0. Follow the steps above to locally overwrite `test-harness/.env`
1. Copy the file portion in the resulting `.env` file starting from `TYPE=...` all the way to the end into `.test-env`
2. Make sure to set `INTERACTIVE_TESTING_ENVIRONMENT=0` and `OVERWRITE_TESTING_ENVIRONMENT=1`
3. Commit these changes and push to github

C.I. should now run against the new harness env variables.

### CODE REVIEW

File: harness_overwrite.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
+import sys
+from typing import Dict, Optional, Tuple
+
+
+def get_env(line) -> Optional[Tuple[str, str, str]]:
```
Comment (tzaffi) [inline harness_overwrite.py:5]:
```suggestion
def get_env(line: str) -> Optional[Tuple[str, str, str]]:
```

File: test-harness.sh [hunk 2]
```diff
@@ -84,6 +84,10 @@ if [[ $OVERWRITE_TESTING_ENVIRONMENT == 1 ]]; then
   cp "$ENV_FILE" "$SDK_TESTING_HARNESS"/.env
 fi
 
+echo "$THIS: INTERACTIVE_TESTING_ENVIRONMENT=$INTERACTIVE_TESTING_ENVIRONMENT"
+if [[ $INTERACTIVE_TESTING_ENVIRONMENT == 1 ]]; then
+  python harness_overwrite.py "$SDK_TESTING_HARNESS"/.env
+fi
```
Comment (tzaffi) [inline test-harness.sh:90]:
```suggestion
# **** BEGIN PYTHON ONLY **** #
echo "$THIS: INTERACTIVE_TESTING_ENVIRONMENT=$INTERACTIVE_TESTING_ENVIRONMENT"
if [[ $INTERACTIVE_TESTING_ENVIRONMENT == 1 ]]; then
  python harness_overwrite.py "$SDK_TESTING_HARNESS"/.env
fi
# **** END PYTHON ONLY **** #
```

File: harness_overwrite.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    print("_" * 50)
+    print("Please provide env variable overrides (skip to keep defaults)")
+    new_env = get_rewrites(env, key_filter=keys)
+
+    print("_" * 50)
+    print("OVERWRITING")
+    for k, v in new_env.items():
+        print(f"new_env[{k}]={v} (PREVIOUS: {env[k]=})")
+
+    overwrite(env_file, new_env)
```
Comment (tzaffi) [inline harness_overwrite.py:81]:
```suggestion
    overwrite(env_file, new_env)
    print("_" * 50)
```

File: harness_overwrite.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
+import sys
+from typing import Dict, Optional, Tuple
+
+
+def get_env(line: str) -> Optional[Tuple[str, str, str]]:
+    if line and line[0] not in (" ", "#") and "=" in line:
+        key, others = line.split("=", maxsplit=1)
+        key = key.strip()
+        if "#" in others:
+            val, extra = others.split("#")
```
Comment (tzaffi) [inline harness_overwrite.py:10]:
```suggestion
            val, extra = others.split("#", maxsplit=1)
```

File: harness_overwrite.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        key = key.strip()
+        if "#" in others:
+            val, extra = others.split("#", maxsplit=1)
+            val = val.strip()
+        else:
+            val, extra = others.strip(), ""
+        return key, val, extra
+
+
+def manualy_parse(env_file: str) -> Dict[str, str]:
```
Comment (tzaffi) [inline harness_overwrite.py:17]: "manually" as opposed with using `dot_env` which I couldn't figure out how to use to read from only one env file and ignore all other env variables.

File: .test-env [hunk 6]
```diff
@@ -13,4 +13,44 @@ REMOVE_LOCAL_FEATURES=0
 
 # WARNING: Be careful when turning on the next variable.
 # In that case you'll need to provide all variables expected by `algorand-sdk-testing`'s `.env`
-OVERWRITE_TESTING_ENVIRONMENT=0
+OVERWRITE_TESTING_ENVIRONMENT=1
+
+# What can ease some of the pain above is to overwrite interactively:
```
Comment (tzaffi) [inline .test-env:18]:
```suggestion
# WARNING: THE FOLLOWING IS FOR LOCAL DEVELOPMENT ONLY
# IF SET TO 1 AND PUSHED TO C.I., THE TEST WILL HANG AND FAIL.
#
# What can ease some of the pain above is to overwrite interactively:
```

File: .test-env [hunk 7]
```diff
@@ -13,4 +13,44 @@ REMOVE_LOCAL_FEATURES=0
...
+INTERACTIVE_TESTING_ENVIRONMENT=0
+
+
```
Comment (tzaffi) [inline .test-env:21]: - [ ] THE FOLLOWING SHOULD BE DELETED BEFORE MERGING

File: .test-env [hunk 8]
```diff
@@ -13,4 +13,47 @@ REMOVE_LOCAL_FEATURES=0
...
+OVERWRITE_TESTING_ENVIRONMENT=1
+
+# WARNING: THE FOLLOWING IS FOR LOCAL DEVELOPMENT ONLY
+# IF SET TO 1 AND PUSHED TO C.I., THE TEST WILL HANG AND FAIL.
+#
+# What can ease some of the pain above is to overwrite interactively:
+INTERACTIVE_TESTING_ENVIRONMENT=0
+
+
+## --- FOR OVERWRITE PURPOSES
```
Comment (tzaffi) [inline .test-env:25]:
```suggestion
## --- FOR OVERWRITE PURPOSES! DELETE ALL THE BELOW!!!!
```

**Comment by bbroder-algo** [comment]
zeph how do we feel about this one?

**Comment by tzaffi** [comment]
> zeph how do we feel about this one?

I think it's kind of cool, but someone else should give a spin to make sure it's worthwhile. I list some pros/cons above, the big con being the future maintenance cost.

**Comment by algochoi** [comment]
I think it is a nice improvement for local testing, and could be worthwhile having it in a long lasting fork or a branch, and reference it in the docs. 

I don't feel as strongly about merging this into `develop` though, for reasons Zeph outlined above. I think having a uniform CI process across all the SDKs seems favorable. If this was implemented/enforced across the four SDKs we manage, I'd be more inclined to merge it in.

**Comment by tzaffi** [comment]
> I think it is a nice improvement for local testing, and could be worthwhile having it in a long lasting fork or a branch, and reference it in the docs.
> 
> I don't feel as strongly about merging this into `develop` though, for reasons Zeph outlined above. I think having a uniform CI process across all the SDKs seems favorable. If this was implemented/enforced across the four SDKs we manage, I'd be more inclined to merge it in.

I'm not inclined to pursue a strong counter-argument recognizing that much of my excitement id due to my personal investment in it. Unless I hear back from more people by the end of the week, I'll close this PR (but will keep a copy in my personal repo for future testing needs).

**Comment by tzaffi** [comment]
I'm closing this PR as there was insufficient interest in pursuing it further. However, I'm keeping [a copy on my personal repo ](https://github.com/tzaffi/py-algorand-sdk/pull/3) for future reference.

**Pull request closed without merging**